{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## EigenSNP-WAS - Gene-level Association Function\n",
    "## Steven D. Shaw"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you would like to use this function, please feel free to do so, and give credit if it is due. I am happy to answer any questions regarding this function, or, if you would like to work together on a project related to gene-level association, please contact me. \n",
    "\n",
    "Email: shawsd@umich.edu\n",
    "\n",
    "Current status: The function is in full working order, and outputs a .txt file with one gene assocation per DV on each line with the following variables, gene name, number of SNPs observed in the gene, number of eigenSNPs within the gene, association statistics (F-value, sum of squares, p-value), and DV (as multiple DVs can be tested at once). To get the function working properly with your own data, ensure that behavioral/ survey/ phenotypic data is formated properly (see comments throughout code) and genetic data has been pre-cleaned in PLINK (e.g., MAF, MIND, etc.). \n",
    "\n",
    "I am in the process of optimizing the speed of the function. Currently, with a dataset of approx. 15k individuals, hg18 gene set (approx. 20k genes), and 4 phenotypes, it takes 4.5 hours to run to completion.\n",
    "\n",
    "If you happen to notice any errors, please do let me know - we all make mistakes and can help eachother overcome and learn from our mistakes :)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pandas_plink import read_plink1_bin\n",
    "from os.path import join\n",
    "from os import path\n",
    "from pandas_plink import get_data_folder\n",
    "from pandas_plink import read_plink\n",
    "import os\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "from datetime import datetime\n",
    "\n",
    "import statsmodels.api as sm\n",
    "import scipy.stats as stats\n",
    "from statsmodels.api import add_constant\n",
    "from sklearn import decomposition\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "import time\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DATA LOADING & MERGING"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genetic Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in genetic data\n",
    "    # QC PLINK vars: --mind 0.05 --geno 0.01 --maf 0.01\n",
    "    # Rename SNPs: --update-map SNP_kgpID2rsID.txt --update-name\n",
    "\n",
    "# Insert name of genetic data file - should be precleaned in PLINK\n",
    "gene_data_filename = ''\n",
    "    \n",
    "(bim, fam, bed) = read_plink(gene_data_filename, verbose=False)\n",
    "bim.set_index('chrom')\n",
    "fam.set_index('i')\n",
    "bim = bim.sort_values(by=['chrom', 'pos'])\n",
    "\n",
    "bim.to_csv(\"bim.csv\")\n",
    "\n",
    "########\n",
    "\n",
    "# Provide a genome gene set list (E.g., 'glist-hg18.txt')\n",
    "gene_set_filename = ''\n",
    "\n",
    "# List of genes in the human genome, and their location (chromosome, start, and end)\n",
    "hg_18 = pd.read_csv(gene_set_filename, \n",
    "                 sep=' ', \n",
    "                 header=None, \n",
    "                 names=[\"chromosome\", \"Start\", \"End\", \"gene\"])\n",
    "\n",
    "# add column that is current position just in case for joining afterwards\n",
    "hg_18['row_num'] = np.arange(len(hg_18))\n",
    "\n",
    "hg_18 = hg_18.sort_values(by=['Start', 'End'])\n",
    "hg_18"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MAP SNPs TO GENES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Reorganize the genetic data (bim) to determine every gene that each SNP is located on\n",
    "# Use gene location information in glist-hg19.txt (dataset: hg_18) to traverse through chromosomes, sequentially,\n",
    "# gene-by-gene and identify which SNPs are in each gene\n",
    "\n",
    "hg_18.set_index('chromosome')\n",
    "\n",
    "snp_to_gene_mapping = []\n",
    "for i in ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', \n",
    "          '11', '12', '13', '14', '15', '16', '17', '18', '19',\n",
    "          '20', '21', '22']:\n",
    "    now = datetime.now()\n",
    "    current_time = now.strftime(\"%H:%M:%S\")\n",
    "    print(\"Current Time =\", current_time)\n",
    "    print(\"chromosome: \", i)\n",
    "    chrom_x = bim.query(\"chrom=='{0}'\".format(i))\n",
    "    hg_x = hg_18.query(\"chromosome=='{0}'\".format(i))\n",
    "    start = hg_x['Start'].min()\n",
    "    end = hg_x['End'].max()\n",
    "    chrom_x = chrom_x[chrom_x[\"pos\"].between(start, end)]\n",
    "    length = len(chrom_x)\n",
    "\n",
    "    j = 0\n",
    "    \n",
    "    for ind, r in hg_x.iterrows():\n",
    "        while chrom_x.iloc[j, 3] < r['Start']:\n",
    "            j += 1\n",
    "            if j >= length:\n",
    "                break\n",
    "        if j >= length:\n",
    "                break\n",
    "        while chrom_x.iloc[j, 3] <= r['End']:\n",
    "            snp_to_gene_mapping.append([chrom_x.iloc[j, 0], chrom_x.iloc[j, 1], chrom_x.iloc[j, 2], \n",
    "                                chrom_x.iloc[j, 3], chrom_x.iloc[j, 4], chrom_x.iloc[j, 5],\n",
    "                                chrom_x.iloc[j, 6], r['Start'], r['End'], r['gene']])\n",
    "            j += 1\n",
    "            if j >= length:\n",
    "                break\n",
    "        if j >= length:\n",
    "                break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Organize traversal into dataframe\n",
    "bim_with_genes = pd.DataFrame(snp_to_gene_mapping, columns = ['chrom' , 'snp', 'cm', 'pos', \n",
    "                                                               'a0', 'a1', 'i','start', 'end', 'gene'])\n",
    "print(bim_with_genes)\n",
    "\n",
    "# Save this file, so you don't need to rerun above code on other computers\n",
    "# bim_with_genes.to_csv(\"bim_with_genes.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load in bim_with_genes for EigenSNP function\n",
    "#bim_with_genes = pd.read_csv('bim_with_genes.csv')\n",
    "#print(bim_with_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## EIGENSNP-WAS FUNCTION"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Required Functions\n",
    "\n",
    "PCA NOTES:\n",
    "\n",
    "Takes in a df where each row is a participant and each column is a SNP. Data input should be minor allele dosage.\n",
    "\n",
    "PCA is run once per gene in main function.\n",
    "\n",
    "It will output a df where each row is a participant and each column contains the eigenvectors from the PCA.\n",
    "\n",
    "The number of PCs is determined by the cummulative variance required set by the 'limit' in the function (typically, greater than 90% variance explained).\n",
    "\n",
    "\n",
    "REGRESION NOTES:\n",
    "\n",
    "Takes in two dfs:\n",
    "\n",
    "i_df : a df where rows are participants and columns are the independent variables\n",
    "\n",
    "t_df : a df where rows are participants and column is the target dependent variable (i.e. BMI)\n",
    "\n",
    "BOTH DATAFRAMES SHOULD HAVE SAME PARTICIPANT ORDER"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##PCA Functions\n",
    "def PC_eigen_vals(gene_df):\n",
    "    X = StandardScaler().fit_transform(gene_df)\n",
    "    limit = 0.95\n",
    "\n",
    "    #perform PCA\n",
    "    pca = decomposition.PCA(n_components=len(gene_df.columns))\n",
    "    #print(pca.fit(X))\n",
    "    pca.fit(X)\n",
    "    X = pca.transform(X)\n",
    "    \n",
    "    #extract variance for each PC\n",
    "    var = pca.explained_variance_ratio_\n",
    "    \n",
    "    # find num PCs\n",
    "    #np.cumsum(var)\n",
    "    \n",
    "    # find num PCs\n",
    "    cum_var = 0\n",
    "    i = 0\n",
    "    while cum_var <= limit and i < len(var):\n",
    "        cum_var += var[i]\n",
    "        i += 1\n",
    "    \n",
    "    # separate eigen vect columns based on limit\n",
    "    X_PCs = X[:,0:i]\n",
    "    \n",
    "    \n",
    "    # add to dictionary\n",
    "    PCs = {}\n",
    "    for i in range(0,X_PCs.shape[1]):\n",
    "        PCs[\"PC{0}\".format(i)] = pd.Series(X[:,i:i+1].transpose()[0])\n",
    "    \n",
    "    # convert to df\n",
    "    PC_df = pd.DataFrame(PCs)\n",
    "    \n",
    "    # Returns pd df\n",
    "    return PC_df\n",
    "\n",
    "## Regression Functions\n",
    "\n",
    "# takes input df and target df\n",
    "def OLS_mod(i_df, t_df):\n",
    "    # set dep and ind variables\n",
    "    X = i_df[[i for i in i_df.columns]]\n",
    "    y = t_df[[i for i in t_df.columns]]\n",
    "\n",
    "    # add constant (intercepts)\n",
    "    X = sm.add_constant(X)\n",
    "\n",
    "    # least squares model\n",
    "    mod = sm.OLS(y, X).fit()\n",
    "    #print(mod.summary())\n",
    "\n",
    "    return mod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#FUNCTION\n",
    "\n",
    "# This loop takes each gene that we observe, obtained only the genetic dosage data that for that gene from the main\n",
    "# genetic dataset, then:\n",
    "    # conducts PCA on SNPs within that gene, and retains eigenvectors from the number of PCs \n",
    "        # that explain >90% variance (i.e., eigenSNPs)\n",
    "    # merges the eigenSNPs with survey/non-genetic data for use with regression\n",
    "    # runs treatment and control regression models and compares these models to get the unique effect of the gene\n",
    "    # outputs the number of SNPs observed in the gene, the number of eigenSNPs, and the p-value\n",
    "\n",
    "#SURVEY DATA\n",
    "    #Survey data must contain continuous dependent variables (variable names inserted into list below),\n",
    "    #participant ID as 'IID', 'Age', and principle components from a genome wide PCA, labeled \n",
    "    #'EV1', 'EV2', 'EV3', 'EV4', etc. \n",
    "    \n",
    "EigenSNPWAS_results = pd.DataFrame(columns=['Gene', 'SNPs', 'EigenSNPs', 'ss_diff', 'f_stat','p_val','DV'])\n",
    "\n",
    "#Chromosome counter so we can stop the loop, for testing\n",
    "chrom_counter = 0\n",
    "\n",
    "bim_with_genes.set_index('chrom')\n",
    "\n",
    "#Get a list of chromosomes in genetic dataset\n",
    "chrom_x = bim_with_genes['chrom'].unique()\n",
    "\n",
    "# List the names of your DVs:\n",
    "DV_list = ['']\n",
    "\n",
    "#Current time\n",
    "current_time = now.strftime(\"%H:%M:%S\")\n",
    "print(\"Current Time =\", current_time)\n",
    "\n",
    "#Start timer\n",
    "start_time = time.time()\n",
    "\n",
    "for j in chrom_x:\n",
    "    \n",
    "    time_chk_chrom_loop = time.time()\n",
    "    \n",
    "#     # Specify how many chromosomes to run - remove to run all\n",
    "#     if chrom_counter == 2:\n",
    "#         break\n",
    "    \n",
    "    print('Chromosome:', j)\n",
    "    \n",
    "    chrom_df = bim_with_genes.query(\"chrom=='{0}'\".format(j))\n",
    "    #Create list of observed genes\n",
    "    chrom_df.set_index('gene')\n",
    "    genes_in_chrom_x = chrom_df['gene'].unique()\n",
    "    \n",
    "    chrom_counter += 1\n",
    "    \n",
    "    #Chromosome counter so we can stop the loop, for testing\n",
    "    gene_counter = 0\n",
    "#t1\n",
    "    time_chk_1 = time.time()\n",
    "    \n",
    "    for i in genes_in_chrom_x:\n",
    "        \n",
    "        time_chk_gene_loop = time.time()\n",
    "        \n",
    "#         # Specify how many genes to run - remove to run all\n",
    "#         if gene_counter == 5:\n",
    "#             break\n",
    "\n",
    "        #Display progress bar\n",
    "            #Change division size for different progress indicators\n",
    "        if gene_counter in list(range(1, len(genes_in_chrom_x), 500)): \n",
    "            print('Gene:', gene_counter, 'of', len(genes_in_chrom_x))\n",
    "            elapsed_time = time.time() - start_time\n",
    "            print(\"Elapsed Time =\", time.strftime(\"%H:%M:%S\", time.gmtime(elapsed_time)))\n",
    "\n",
    "        # Get the SNP data from bed file\n",
    "        gene_df = chrom_df.query(\"gene=='{0}'\".format(i))\n",
    "        x = bed[gene_df.i.values, :].compute() #x contains only dosage data for SNPs in current gene\n",
    "\n",
    "        # transpose so that data are formatted correctly (columns are SNPs, rows are participants)\n",
    "        x_trans = x.transpose()\n",
    "        #print(x_trans)\n",
    "\n",
    "        # Get SNP names, create dataframe with SNP labels (as columns) for use in PCA\n",
    "        snps = gene_df['snp']\n",
    "        SNP_data_for_PC = pd.DataFrame(x_trans, columns = snps)\n",
    "        #print('with NaN:', SNP_data_for_PC)\n",
    "\n",
    "        #Replace NaNs with mean for each column\n",
    "        SNP_data_for_PC.fillna(SNP_data_for_PC.mean(), axis=0, inplace=True)\n",
    "\n",
    "        #There are some strange columns of all NaNs, remove these first\n",
    "        SNP_data_for_PC.dropna(axis=0, how='all', inplace=True)\n",
    "\n",
    "        #finally, if any NAs remain within a row, drop that row\n",
    "            #very few of these cases, but needed to keep function running (e.g., CSMD1 in chr8, has random missingness)\n",
    "        SNP_data_for_PC.dropna(axis=1, how='any', inplace=True)\n",
    "        #print(SNP_data_for_PC.shape)\n",
    "        #print('without NaN:', SNP_data_for_PC)\n",
    "\n",
    "        #Count number of SNPs\n",
    "        SNP_count = len(SNP_data_for_PC.columns)\n",
    "        #print(i)\n",
    "        #print(SNP_count)\n",
    "#t2 \n",
    "        time_chk_2 = time.time()\n",
    "        \n",
    "        try:\n",
    "            if SNP_count == 0:\n",
    "                #EigenSNPWAS.append([i, 0, np.nan, np.nan])\n",
    "                EigenSNPWAS_results = EigenSNPWAS_results.append({'Gene': i,\n",
    "                                'SNPs': 0,\n",
    "                                'EigenSNPs': np.nan,\n",
    "                                'ss_diff': np.nan,\n",
    "                                'f_stat': np.nan,\n",
    "                                'p_val': np.nan,\n",
    "                                'DV': np.nan},                                \n",
    "                                ignore_index=True)\n",
    "            else:\n",
    "                if SNP_count == 2:\n",
    "                    PCA = SNP_data_for_PC\n",
    "                else:\n",
    "                    #Conduct PCA - output contains dataframe with EigenSNPs (cut to >90% threshold)\n",
    "                    PCA = PC_eigen_vals(SNP_data_for_PC)\n",
    "                    #print(PCA)\n",
    "\n",
    "                #Count number of EigenSNPs\n",
    "                PC_Eigen_count = len(PCA.columns)\n",
    "                #print(PC_Eigen_count)\n",
    "\n",
    "                # Take output of PCA - eigenSNPs - and merge with survey/non-genetic data\n",
    "                #Get ID codes for gene data - use key values (i) from query with gene name\n",
    "                    #gene_df contains only the SNPs we are interested in\n",
    "\n",
    "                #Append gene IDs to PCA output\n",
    "                PCA['IID'] = pd.to_numeric(fam['iid'], errors=\"coerce\")\n",
    "                # new code - Add Sex variable from fam file\n",
    "                PCA['Sex'] = pd.to_numeric(fam['gender'], errors=\"coerce\")\n",
    "                #print('fam iid_type:', PCA['IID'].dtypes)\n",
    "                #print(PCA)\n",
    "\n",
    "                #merge eigenSNPs with dataset: DATA_survey \n",
    "                PCA = PCA[PCA['IID'].notna()]\n",
    "#t3                \n",
    "                time_chk_3 = time.time()\n",
    "                \n",
    "                for DV in DV_list:\n",
    "                    DATA_survey_DV = DATA_survey[[DV, 'IID', 'Age', 'EV1', 'EV2', 'EV3', 'EV4']]\n",
    "                    DATA_all_for_gene = pd.merge(DATA_survey_DV, PCA, left_on='IID', right_on='IID', how='inner')\n",
    "                    #print(list(DATA_all_for_gene.columns)) #View columns in dataset\n",
    "\n",
    "                    #Drop the IID column, as it is no longer needed\n",
    "                    DATA_all_for_gene.drop('IID', axis=1, inplace=True)\n",
    "                    #Remove any rows with missing values\n",
    "                    DATA_all_for_gene.dropna(inplace=True)\n",
    "                    DATA_DV_only = pd.DataFrame(DATA_all_for_gene[DV])\n",
    "\n",
    "                    #Create treatment dataset\n",
    "                    DATA_treatment_model = DATA_all_for_gene.drop(DV, axis=1)\n",
    "                    #print(DATA_treatment_model)\n",
    "\n",
    "                    #Create control dataset\n",
    "                    DATA_control_model = DATA_all_for_gene[['Age', 'Sex', 'EV1', 'EV2', 'EV3', 'EV4']]\n",
    "                    #print(DATA_control_model)\n",
    "\n",
    "                    # TREATMENT Model\n",
    "                    treat_mod = OLS_mod(DATA_treatment_model, DATA_DV_only)     ## controls + EigenSNPs\n",
    "                    #treat_mod.summary()\n",
    "\n",
    "                    # CONTROL Model\n",
    "                    cont_mod = OLS_mod(DATA_control_model, DATA_DV_only)   ## controls only\n",
    "                    #cont_mod.summary()\n",
    "\n",
    "                    #Conduct ANOVA to compare TREATMENT and CONTROL models\n",
    "                    aov_table = sm.stats.anova_lm(cont_mod, treat_mod, typ=1)\n",
    "                    #print(aov_table)\n",
    "\n",
    "                    # Extract stats from model (F and p-val)\n",
    "                    gene_ss_diff = aov_table['ss_diff'][1]\n",
    "                    gene_fstat = aov_table['F'][1]\n",
    "                    gene_pval = aov_table['Pr(>F)'][1]\n",
    "                    #print('f:', gene_fstat, 'p:', gene_pval)\n",
    "\n",
    "                    #print('Gene:', i, 'SNPs:', SNP_count, 'EigenSNPs:', PC_Eigen_count, 'p_val:', gene_pval)\n",
    "\n",
    "                    #EigenSNPWAS.append([i, SNP_count, PC_Eigen_count, gene_pval])\n",
    "                    EigenSNPWAS_results = EigenSNPWAS_results.append({'Gene': i,\n",
    "                                                                    'SNPs': SNP_count,\n",
    "                                                                    'EigenSNPs': PC_Eigen_count,\n",
    "                                                                    'ss_diff': gene_ss_diff,\n",
    "                                                                    'f_stat': gene_fstat,\n",
    "                                                                    'p_val': gene_pval,\n",
    "                                                                    'DV': DV},\n",
    "                                                                    ignore_index=True)\n",
    "        except ValueError:  #happens once in chromosome 8 (dealt with, but will leave try/except just in case)\n",
    "            EigenSNPWAS_results = EigenSNPWAS_results.append({'Gene': i,\n",
    "                                                            'SNPs': SNP_count,\n",
    "                                                            'EigenSNPs': np.nan,\n",
    "                                                            'ss_diff': np.nan,\n",
    "                                                            'f_stat': np.nan,\n",
    "                                                            'p_val': np.nan,\n",
    "                                                            'DV': np.nan},                                \n",
    "                                                            ignore_index=True)\n",
    "#t4\n",
    "        time_chk_4 = time.time()\n",
    "        gene_counter += 1\n",
    "#         print('Time_checks:', '1:', time_chk_1-time_chk_chrom_loop, '2:', time_chk_2-time_chk_gene_loop,\n",
    "#              '3:', time_chk_3-time_chk_2, '4:', time_chk_4-time_chk_3)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
